src/rnp/rnp.c

```diff
@@ -304,181 +308,209 @@ rnp_on_signatures(pgp_parse_handler_t *handler, pgp_signature_info_t *sigs, int
-    rnp_ctx_t ctx = {0};
-    // TODO: Probably something smarter should be done here
-    repgp_io_t *io = repgp_create_io();
-
-    if (io == NULL) {
-        RNP_LOG("Allocation failed");
-        return false;
-    }
+    int         cmd;
+    const char *fname;
 
+    /* some rnp_t setup */
     if (rnp_cfg_getstr(cfg, CFG_PASSWD)) {
         rnp->password_provider.callback = rnp_password_provider_string;
         rnp->password_provider.userdata = (void *) rnp_cfg_getstr(cfg, CFG_PASSWD);
     }
+    rnp->pswdtries = rnp_cfg_get_pswdtries(cfg);
 
-    /* operation context initialization: writing all additional parameters */
-    rnp_ctx_init(&ctx, rnp);
-    ctx.armor = rnp_cfg_getint(cfg, CFG_ARMOR);
-    ctx.overwrite = rnp_cfg_getbool(cfg, CFG_OVERWRITE);
-    if (f) {
-        ctx.filename = strdup(rnp_filename(f));
-        ctx.filemtime = rnp_filemtime(f);
+    /* operation context initialization */
+    rnp_ctx_init(ctx, rnp);
+    ctx->armor = rnp_cfg_getint(cfg, CFG_ARMOR);
+    ctx->overwrite = rnp_cfg_getbool(cfg, CFG_OVERWRITE);
+    if ((fname = rnp_cfg_getstr(cfg, CFG_INFILE))) {
+        ctx->filename = strdup(rnp_filename(fname));
+        ctx->filemtime = rnp_filemtime(fname);
     }
-    rnp->pswdtries = rnp_cfg_get_pswdtries(cfg);
 
-    switch (cmd) {
-    case CMD_CLEARSIGN:
-    case CMD_SIGN:
-        ctx.halg = pgp_str_to_hash_alg(rnp_cfg_getstr(cfg, CFG_HASH));
+    /* get the ongoing command. OpenPGP commands are only ENCRYPT/ENCRYPT_SIGN/SIGN/DECRYPT */
+    cmd = rnp_cfg_getint(cfg, CFG_COMMAND);
 
-        if (ctx.halg == PGP_HASH_UNKNOWN) {
-            fprintf(stderr, "Unknown hash algorithm: %s\n", rnp_cfg_getstr(cfg, CFG_HASH));
-            ret = false;
-            break;
-        }
+    /* options used for signing */
+    if (cmd == CMD_PROTECT) {
+        ctx->zalg = rnp_cfg_getint(cfg, CFG_ZALG);
+        ctx->zlevel = rnp_cfg_getint(cfg, CFG_ZLEVEL);
 
-        rnp_cfg_copylist_str(cfg, &ctx.signers, CFG_SIGNERS);
+        /* setting signing parameters if needed */
+        if (rnp_cfg_getbool(cfg, CFG_SIGN_NEEDED)) {
+            ctx->halg = pgp_str_to_hash_alg(rnp_cfg_getstr(cfg, CFG_HASH));
 
-        if (!list_length(ctx.signers)) {
-            if (!rnp->defkey) {
-                fprintf(stderr, "No userid or default key for signing\n");
-                ret = false;
-                break;
+            if (ctx->halg == PGP_HASH_UNKNOWN) {
+                fprintf(stderr, "Unknown hash algorithm: %s\n", rnp_cfg_getstr(cfg, CFG_HASH));
+                return false;
             }
 
-            if (!list_append(&ctx.signers, rnp->defkey, strlen(rnp->defkey) + 1)) {
-                RNP_LOG("allocation failed");
-                ret = false;
-                break;
+            rnp_cfg_copylist_str(cfg, &ctx->signers, CFG_SIGNERS);
+
+            if (!list_length(ctx->signers)) {
+                if (!rnp->defkey) {
+                    fprintf(stderr, "No userid or default key for signing\n");
```

Comment:
RNP_LOG

Reply:
This is actually not 'technical' logging but message to user. Later on we should separate it with some specific function, responsible for interaction with user.